pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/SchedulerManager.java

```diff
@@ -403,6 +548,82 @@ private void scheduleCompaction(ScheduledExecutorService executor, long schedule
         }
     }
 
+    @VisibleForTesting
+    List<Assignment> getCurrentPostDrainAssignments() {
+        return currentPostDrainAssignments;
+    }
+
+    @VisibleForTesting
+    void clearCurrentPostDrainAssignments() {
+        currentPostDrainAssignments = null;
+    }
+
+    @VisibleForTesting
+    List<Assignment> invokeDrain(String workerId) {
+
+        long startTime = System.nanoTime();
+
+        Set<String> availableWorkers = getCurrentAvailableWorkers();
+
+        // workerIdToAssignments is a map of the form {workerId : {FullyQualifiedInstanceId : Assignment}}
+        Map<String, Map<String, Assignment>> workerIdToAssignments = functionRuntimeManager.getCurrentAssignments();
+
+        // initialize stats collection
+        SchedulerStats schedulerStats = new SchedulerStats(workerIdToAssignments, availableWorkers);
+
+        boolean drainSuccessful = false;
+        List<Assignment> postDrainAssignments = null;
+
+        try {
+            // The worker was checked to be among the available ones, by the caller.
+            Preconditions.checkState(availableWorkers.contains(workerId));
```

Comment:
What if the worker is restarting during when this happens?  The worker can temporarily have disappeared from the the membership. Shouldn't we proceed to removing the assignments from the worker and still mark it as in progress of being drained.

Reply:
Sounds fair; I will remove the precondition check.